File: chatter/scripts/ops/validate_protocol_schemas.mjs
#!/usr/bin/env node
import { promises as fs } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import Ajv from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const repoRoot = path.resolve(__dirname, "..", "..");

const schemaMap = {
  StreamContext: path.join(
    repoRoot,
    "packages/protocol/jsonschema/stream_context.schema.json"
  ),
  ChatMessage: path.join(
    repoRoot,
    "packages/protocol/jsonschema/chat_message.schema.json"
  ),
  TrendsSnapshot: path.join(
    repoRoot,
    "packages/protocol/jsonschema/trends_snapshot.schema.json"
  ),
};

const parseArgs = () => {
  const args = process.argv.slice(2);
  const options = { version: "1.0.0", only: null };
  for (let i = 0; i < args.length; i += 1) {
    if (args[i] === "--only" && args[i + 1]) {
      options.only = args[i + 1];
      i += 1;
    } else if (args[i] === "--version" && args[i + 1]) {
      options.version = args[i + 1];
      i += 1;
    }
  }
  return options;
};

const loadJson = async (filePath) => {
  const content = await fs.readFile(filePath, "utf-8");
  return JSON.parse(content);
};

const listJsonFiles = async (directory) => {
  const entries = await fs.readdir(directory).catch(() => []);
  return entries
    .filter((entry) => entry.endsWith(".json"))
    .sort()
    .map((entry) => path.join(directory, entry));
};

const validateSchema = async (name, schemaPath, ajv) => {
  console.log(`Checking schema for ${name}: ${schemaPath}`);
  try {
    const schema = await loadJson(schemaPath);
    ajv.compile(schema);
    console.log("  [OK] Schema is valid and compiled");
    return schema;
  } catch (error) {
    console.error(`  [FAIL] Schema error for ${name}: ${error.message}`);
    return null;
  }
};

const validateFixtures = async (name, version, schema, ajv) => {
  const validator = ajv.compile(schema);
  const baseDir = path.join(repoRoot, "data/schemas", name, version);
  const validDir = path.join(baseDir, "valid");
  const invalidDir = path.join(baseDir, "invalid");

  let passed = 0;
  let failed = 0;

  const validDirExists = await fs
    .access(baseDir)
    .then(() => true)
    .catch(() => false);

  if (!validDirExists) {
    console.warn(
      `[WARN] Fixture directory missing for ${name} ${version}: ${baseDir}`
    );
    return { passed, failed };
  }

  const scenarios = [
    { label: "valid", directory: validDir, expectValid: true },
    { label: "invalid", directory: invalidDir, expectValid: false },
  ];

  for (const { label, directory, expectValid } of scenarios) {
    const files = await listJsonFiles(directory);
    if (files.length === 0) {
      console.warn(
        `[WARN] ${label} fixtures missing or empty for ${name} ${version}: ${directory}`
      );
      continue;
    }

    console.log(`Validating ${label} fixtures for ${name} ${version}: ${directory}`);
    for (const filePath of files) {
      try {
        const data = await loadJson(filePath);
        const isValid = validator(data);
        if (isValid && expectValid) {
          console.log(`  [OK] ${filePath}`);
          passed += 1;
        } else if (!isValid && !expectValid) {
          console.log(
            `  [OK] ${filePath}: correctly failed validation (${ajv.errorsText(validator.errors)})`
          );
          passed += 1;
        } else if (isValid && !expectValid) {
          console.error(
            `  [FAIL] ${filePath}: expected failure but validation succeeded`
          );
          failed += 1;
        } else {
          console.error(
            `  [FAIL] ${filePath}: validation failed (${ajv.errorsText(validator.errors)})`
          );
          failed += 1;
        }
      } catch (error) {
        console.error(`  [FAIL] ${filePath}: unable to validate (${error.message})`);
        failed += 1;
      }
    }
  }

  return { passed, failed };
};

const main = async () => {
  const { only, version } = parseArgs();
  const ajv = new Ajv({ strict: false });
  addFormats(ajv);

  const targets = only ? [only] : Object.keys(schemaMap).sort();

  let totalPassed = 0;
  let totalFailed = 0;

  for (const name of targets) {
    const schemaPath = schemaMap[name];
    if (!schemaPath) {
      console.error(`[FAIL] Unknown schema: ${name}`);
      totalFailed += 1;
      continue;
    }
    const schemaExists = await fs
      .access(schemaPath)
      .then(() => true)
      .catch(() => false);

    if (!schemaExists) {
      console.error(`[FAIL] Missing schema for ${name}: ${schemaPath}`);
      totalFailed += 1;
      continue;
    }

    const schema = await validateSchema(name, schemaPath, ajv);
    if (!schema) {
      totalFailed += 1;
      continue;
    }

    const { passed, failed } = await validateFixtures(name, version, schema, ajv);
    totalPassed += passed;
    totalFailed += failed;
  }

  console.log("\nSummary:");
  console.log(`  Passed: ${totalPassed}`);
  console.log(`  Failed: ${totalFailed}`);

  if (totalFailed > 0) {
    process.exitCode = 1;
  }
};

main();

File: chatter/scripts/ops/validate_protocol_schemas.py
#!/usr/bin/env python
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Dict, Iterable, Tuple

from jsonschema import Draft202012Validator


SchemaConfig = Dict[str, Path]


def load_json(path: Path) -> object:
    with path.open("r", encoding="utf-8") as f:
        return json.load(f)


def gather_paths(root: Path) -> Tuple[Path, Path]:
    valid_dir = root / "valid"
    invalid_dir = root / "invalid"
    return valid_dir, invalid_dir


def validate_schema(name: str, schema_path: Path) -> bool:
    print(f"Checking schema for {name}: {schema_path}")
    if not schema_path.exists():
        print(f"  [FAIL] Schema file missing: {schema_path}")
        return False

    try:
        schema = load_json(schema_path)
    except Exception as exc:  # noqa: BLE001
        print(f"  [FAIL] Could not load schema: {exc}")
        return False

    try:
        Draft202012Validator.check_schema(schema)
        print("  [OK] Schema is valid against Draft 2020-12 metaschema")
        return True
    except Exception as exc:  # noqa: BLE001
        print(f"  [FAIL] Schema validation error: {exc}")
        return False


def validate_fixture(
    fixture_path: Path, validator: Draft202012Validator, expect_valid: bool
) -> bool:
    try:
        instance = load_json(fixture_path)
    except Exception as exc:  # noqa: BLE001
        print(f"  [FAIL] {fixture_path}: unable to load JSON ({exc})")
        return False

    try:
        validator.validate(instance)
        if expect_valid:
            print(f"  [OK] {fixture_path}")
            return True
        print(f"  [FAIL] {fixture_path}: expected validation failure but passed")
        return False
    except Exception as exc:  # noqa: BLE001
        if expect_valid:
            print(f"  [FAIL] {fixture_path}: validation failed ({exc})")
            return False
        print(f"  [OK] {fixture_path}: correctly failed validation ({exc})")
        return True


def validate_fixtures(
    name: str, version: str, schema_path: Path, base_fixture_dir: Path
) -> Tuple[int, int]:
    if not schema_path.exists():
        print(f"[FAIL] Schema missing for {name}: {schema_path}")
        return 0, 1

    try:
        schema = load_json(schema_path)
    except Exception as exc:  # noqa: BLE001
        print(f"[FAIL] Could not load schema for {name}: {exc}")
        return 0, 1

    validator = Draft202012Validator(schema)
    valid_dir, invalid_dir = gather_paths(base_fixture_dir)

    passed = 0
    failed = 0

    if not base_fixture_dir.exists():
        print(f"[WARN] Fixture directory missing for {name} {version}: {base_fixture_dir}")
        return passed, failed

    for label, directory, expect_valid in (
        ("valid", valid_dir, True),
        ("invalid", invalid_dir, False),
    ):
        if not directory.exists():
            print(f"[WARN] Missing {label} fixtures for {name} {version}: {directory}")
            continue
        print(f"Validating {label} fixtures for {name} {version}: {directory}")
        for path in sorted(directory.glob("*.json")):
            if validate_fixture(path, validator, expect_valid):
                passed += 1
            else:
                failed += 1
        if not any(directory.glob("*.json")):
            print(f"  [WARN] No {label} fixtures found in {directory}")

    return passed, failed


def build_schema_map(repo_root: Path) -> SchemaConfig:
    return {
        "StreamContext": repo_root / "packages/protocol/jsonschema/stream_context.schema.json",
        "ChatMessage": repo_root / "packages/protocol/jsonschema/chat_message.schema.json",
        "TrendsSnapshot": repo_root / "packages/protocol/jsonschema/trends_snapshot.schema.json",
        "StreamFrame": repo_root / "packages/protocol/jsonschema/stream_frame.v1.schema.json",
        "StreamTranscriptSegment": repo_root / "packages/protocol/jsonschema/stream_transcript_segment.v1.schema.json",
        "StreamObservation": repo_root / "packages/protocol/jsonschema/stream_observation.v1.schema.json",
    }


def main(argv: Iterable[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Validate protocol schemas and fixtures")
    parser.add_argument(
        "--only",
        choices=[
            "StreamContext",
            "ChatMessage",
            "TrendsSnapshot",
            "StreamFrame",
            "StreamTranscriptSegment",
            "StreamObservation",
        ],
        help="Validate a single schema by name",
    )
    parser.add_argument(
        "--version",
        default="1.0.0",
        help="Version directory to validate (default: 1.0.0)",
    )

    args = parser.parse_args(list(argv) if argv is not None else None)

    repo_root = Path(__file__).resolve().parents[2]
    schema_map = build_schema_map(repo_root)

    targets = [args.only] if args.only else sorted(schema_map.keys())

    total_passed = 0
    total_failed = 0

    for name in targets:
        schema_path = schema_map[name]
        fixtures_dir = repo_root / "data/schemas" / name / args.version

        schema_ok = validate_schema(name, schema_path)
        if not schema_ok:
            total_failed += 1
            continue

        passed, failed = validate_fixtures(name, args.version, schema_path, fixtures_dir)
        total_passed += passed
        total_failed += failed

    print("\nSummary:")
    print(f"  Passed: {total_passed}")
    print(f"  Failed: {total_failed}")

    return 0 if total_failed == 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())
